addons/cloud_storage_azure/models/cloud_storage_azure.py

```diff
@@ -0,0 +1,103 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import re
+from datetime import datetime, timedelta, timezone
+from urllib.parse import unquote, quote
+
+from odoo import models, _
+from odoo.exceptions import ValidationError
+
+from ..utils.cloud_storage_azure_utils import generate_blob_sas, get_user_delegation_key, ClientAuthenticationError
+
+
+class CloudStorageAzure:
+    _url_pattern = re.compile(r'https://(?P<account_name>[\w]+).blob.core.windows.net/(?P<container_name>[\w]+)/(?P<blob_name>[^?]+)')
+    _upload_url_time_to_expiry = 300  # 300 seconds
+    _download_url_time_to_expiry = 300  # 300 seconds
+    __user_delegation_keys = {}  # {db_name, (config, user_delegation_key or exception)}
+
+    @classmethod
+    def get_info_from_url(cls, url):
+        match = cls._url_pattern.match(url)
+        if not match:
+            raise ValidationError(_('%s is not a valid Azure Blob Storage URL.', url))
+        return {
+            'account_name': match['account_name'],
+            'container_name': match['container_name'],
+            'blob_name': unquote(match['blob_name']),
+        }
+
+    @classmethod
+    def get_user_delegation_key(cls, env):
+        """ get a not expired UserDelegationKey for the latest token """
+        cached_config, cached_user_delegation_key = cls.__user_delegation_keys.get(env.registry.db_name, (None, None))
+        db_config = env['res.config.settings']._get_cloud_storage_configuration()
```

Comment:
Since the ormcache is invalidated very frequently in production, this method double checks the ormcached `_get_cloud_storage_configuration()` to avoid slow requests to azure server when the real config is not changed.